### DISCUSSION TITLE
RFC: More agressive DCE & TreeShaking

### DISCUSSION THREAD

**Discussion by cdaringe** [vercel/next.js#33078] [Help]
# Problem

## Short

- no tree shaking on `getInitialProps`
- no tree shaking, post terser, on unused imports

## Complete

1. We use [getInitialProps](https://nextjs.org/docs/api-reference/data-fetching/getInitialProps), and thus do not participate in next's [DCE process](https://next-code-elimination.vercel.app/s/MwVo1ERGq), but we desire to.
    - `getServerSide` props provides similar functionality, but we do _not want_ to use our server after first SSR. `getInitialProps` lets us SSR, then transition to a fully client-side app after first-render. For this reason, we believe that `getInitialProps` should continue to be treated like a first-class citizen, versus being recommended against in the next docs.
1. Unused, side-effect free imports are not tree shaken in client builds

**Desired target state**:

```jsx
// pages/_app.js
export { default } from '../libs/app-page/index.js';

// libs/app-page/index.js
import { foo } from "./bar";
// ^ Full import should be tree-shaken or DCE'd from the client side bundle,
// but this module is bundled instead.

function MyApp() {
  return <App />
}

if (typeof window === "undefined") {
  MyApp.getInitialProps = () => {
    foo();
    return {};
  };
}

export default MyApp;
```

Instead, webpack + terser strip the `if () ...` statement only, leaving the import, yielding something equivalent to:

```js
import { foo } from "./bar"; // still here :|
function MyApp() {
  return <App />
}
export default MyApp;
```

# Demonstration

- [this next-code-elimination URL](https://next-code-elimination.vercel.app/s/Wp-xJQSy_)
- [MVP demo](https://github.com/cdaringe/next-weak-shaking)
- [MVP demo on sideEffects: false](https://github.com/cdaringe/next-weak-shaking-side-effects)

# Discussion

- [Terser won't drop the module](https://github.com/terser/terser/issues/168#issuecomment-440362137)
- `sideEffects: "false"` in our next.js root doesn't have an impact in dropping the `'./bar'` module shown above (or the modules in the demos)

For max performance per ENV, we have various uses of code deploying `typeof window === 'undefined'` conditionals. Sometimes, we swap implementations with server vs browser implementations. Other times, we extend a control flow or just execute a side effect for that environment. In both cases, we expect those conditionals to be erased in opposing envs, and they are! When we started our project ~2 years ago, initial tests showed these DCE/erasures successful, and we believed that the associated imports were erased as well. We've recently realized "hey, a lot of this code we thought should have been shaken, turns out is in the browser bundle".

We'd like to discuss what we can do to allow for shaking/erasure of pure imports.

Outside of `getInitialProps`, sometimes we have valid use cases as well to access server only resources in react code:

```js
import cache from 'some/async-localstorage-cache';
const IS_SERVER = typeof window === 'undefined'

function Component() {
  const apiClient = IS_SERVER ? cache.client : createBrowserClient();
  return <MyApiContext client={client} />;
}
```

We could just _create_ the resource again, but that's wasteful, and we care about micro and macro optimizations. This is just a single example after all :)

**Comment by jescalan** [comment]
Hi there @cdaringe! I spoke with the team about this a bit and unfortunately its not something we are able to prioritize internally, though as always we would, of course, welcome contributions if anyone is interested.

This would be very difficult to accomplish and have many edge cases, and we are currently heavily invested in a couple of projects that we think will solve the same root problem in a more efficient manner -- namely, some improvements to routing within next.js (we are planning to put out an RFC with more detail here in the next few weeks), and react server component support.

Sorry I didn't have a more positive answer for you here and I hope that over the next few months we can bring you some new features that will greatly alleviate this pain for you all 😀

**Reply by cdaringe** [reply]
Hi @jescalan,

Thanks for writing.

Yes, we'd be interested in working on it. If someone could help enumerate the cases in play and confirm that such patches would be welcomed, we could budget time to work on this. We haven't seen any technical discussion on this matter yet, other than "use getServerSideProps" in an email, which does not consider or address our usage of getInitialProps discussed above. We think a config option could allow opt-in participation in next's pre-existing DCE & tree shaking, along with what is already done for the other data fetching methods.

I also recall that internally, next.js only applies optimizations to source in the `pages/*` dir, but our pages actually import page components from something akin to `@our-scope/my-page`, then re-export the components from library space. We do this because as we host multiple applications that _re-use_ page components. Thus, some consideration/support on getting tree-shaking to work more flexibly than just on source files matching `pages/*` would be helpful as well. This topic applies even if we were using `getServerSideProps` vs `getInitialProps`.

Thx for the consideration

**Reply by jescalan** [reply]
Of course! I think it sounds like we should sync up and come to terms with the use case here and what's needed. I'll make sure that we get a call set up and can have this cleared up 😁
